Include tinyformat as a subtree

Empact opened the pull request:
This adds the existing c42f/tinyformat@689695c of tinyformat as a subtree, and extracts the modifications that had been made to a utilstrprintf.h wrapper file which is included instead of tinyformat.h directly.

The benefits being that this should be easier to verifiably update and clearer what changes are being introduced, while making it easier to identify tinyformat.h as a dependency rather than a project file, and apply special treatment relative to that fact.

This re-opens #13842

laanwj commented:
Concept ACK
I didn't do this initially because I didn't expect tinyformat to be an actually maintained library (as well as not really understanding submodules very well), but in retrospect this is better.

scravy commented:
utACK – LGTM

laanwj commented:
labeling this for after the 0.18 branch-off, I don't think this makes sense to do last-minute for 0.17

Empact commented:
@laanwj Agreed

Empact commented:
Repeating @MarcoFalke's concern here for visibility:
"Seems like a lot of files to pull in just to make it a subtree, no? Imo tinyformat is really meant to be a single file just to be copied into the project."
https://github.com/bitcoin/bitcoin/pull/13842#issuecomment-409732591

I have a mild preference for automatic checks of consistency with upstream, but #13846 exists as an alternative.

ajtowns reviewed (commented):
Concept ACK

laanwj commented:
> Seems like a lot of files to pull in just to make it a subtree, no? Imo tinyformat is really meant to be a single file just to be copied into the project."

After seeing the number of files added here I think I'm starting to agree with @MarcoFalke.

Empact commented:
Here are the files included in the subtree - 10 in total: https://github.com/Empact/bitcoin/tree/tinyformat-subtree/src/tinyformat

The other changed files relate mostly to redirecting tinyformat.h includes to utilstrprintf.h.

Empact committed:
Drop src/tinyformat.h in favor of the subtree src/tinyformat/

Introduce modifications to tinyformat via src/tinyformat.h

This includes changes in:
695041e4952ea40e0 - util: Update tinyformat
1b8fd35aadfad6a1e - Make tinyformat errors raise an exception instead of assert()ing
9b6d4c5cdc1ad7b12 - Move strprintf define to tinyformat.h
6e5fd003e04b81115 - Move `*Version()` functions to version.h/cpp
9eaa0afa6ec5d3dd0 - tinyformat: force USE_VARIADIC_TEMPLATES
b651270cd6bfdd6d7 - util: Throw tinyformat::format_error on formatting error

This does not include changes below, and it protects against future unintended
modification of upstream code via git-subtree-check:
64fb0ac016c7fd01c - Declare single-argument (non-converting) constructors "explicit"
4d9b4256d89d1f7c6 - Fix typos

Note I excluded only the tinyformat cpp from the boost lint checks out of an
abundance of caution - these files are not actually built into bitcoin so are
not at risk of pulling in unwanted dependencies, whereas the header, and other
dependencies might be.

ajtowns commented:
The two PRs is a bit of a pain. I think I have a mild preference for the 10-files-but-automatic-checks over 2-files-but-manual-checks, but am not really bothered either way.

The 46-files-changed is mostly just churn to pull out the customisations into utilstrprintf.h. You could avoid that by renaming utilstrprintf.h back to tinyformat.h (which then includes tinyformat/tinyformat.h). If I'm counting right, doing that gets it down to 24 changes: 10 src/tinyformat/, src/tinyformat.h, dev-notes.md, copyright_header.py, Makefile.am, src/Makefile.am, 6 lint files, travis.yml, and src/primitives/block.{h,cpp}. The block.{h,cpp} changes are just to directly include some standard headers so could also be skipped.

ajtowns commented on src/Makefile.am:
```diff
@@ -182,6 +182,7 @@ BITCOIN_CORE_H = \
   util.h \
   utilmemory.h \
   utilmoneystr.h \
+  utilstrprintf.h \
```
Is this change needed/useful?

Empact commented:
One option is to split this into 2 prs: the extraction of utilstrprintf.h and the move to a subdir (which would be pretty trivial at that point).

ken2812221 commented:
Agree with @ajtowns, rename `utilstrprintf.h` to `tinyformat.h`

Empact committed:
Merge commit 'ddbf92320a97efa36994f42144960094e3bd10e6' as 'src/tinyformat'

Empact committed:
Squashed 'src/tinyformat/' content from commit 689695cf5

git-subtree-dir: src/tinyformat
git-subtree-split: 689695cf58700e6defe3741829564cd682d5ae57

laanwj commented:
Closing (see https://github.com/bitcoin/bitcoin/pull/13846#issuecomment-417666596)
